File: src/transport.ts
/**
 * HTTP Transport for Raindrop
 * Fire-and-forget with buffering and retry
 */

import type { TraceData, FeedbackOptions, UserTraits } from './types.js';

interface TransportConfig {
  apiKey: string;
  baseUrl: string;
  debug: boolean;
  disabled: boolean;
}

interface QueuedEvent {
  type: 'trace' | 'feedback' | 'identify';
  data: unknown;
  timestamp: number;
}

export class Transport {
  private config: TransportConfig;
  private queue: QueuedEvent[] = [];
  private flushTimeout: ReturnType<typeof setTimeout> | null = null;
  private readonly FLUSH_INTERVAL = 1000; // 1 second
  private readonly MAX_QUEUE_SIZE = 100;
  private readonly MAX_RETRIES = 3;

  constructor(config: TransportConfig) {
    this.config = config;
  }

  /**
   * Send a trace event (fire-and-forget)
   */
  sendTrace(trace: TraceData): void {
    if (this.config.disabled) return;

    this.enqueue({
      type: 'trace',
      data: this.formatTrace(trace),
      timestamp: Date.now(),
    });
  }

  /**
   * Send feedback/signal event
   */
  sendFeedback(traceId: string, feedback: FeedbackOptions): void {
    if (this.config.disabled) return;

    this.enqueue({
      type: 'feedback',
      data: {
        event_id: traceId,
        signal_name: feedback.type || (feedback.score !== undefined && feedback.score >= 0.5 ? 'positive' : 'negative'),
        sentiment: feedback.score !== undefined
          ? (feedback.score >= 0.5 ? 'POSITIVE' : 'NEGATIVE')
          : (feedback.type === 'thumbs_up' ? 'POSITIVE' : 'NEGATIVE'),
        signal_type: feedback.signalType || 'default',
        timestamp: feedback.timestamp || new Date().toISOString(),
        ...(feedback.attachmentId && { attachment_id: feedback.attachmentId }),
        properties: {
          score: feedback.score,
          comment: feedback.comment,
          ...feedback.properties,
        },
      },
      timestamp: Date.now(),
    });
  }

  /**
   * Send user identification
   */
  sendIdentify(userId: string, traits: UserTraits): void {
    if (this.config.disabled) return;

    this.enqueue({
      type: 'identify',
      data: {
        user_id: userId,
        traits,
      },
      timestamp: Date.now(),
    });
  }

  /**
   * Format trace data for API
   */
  private formatTrace(trace: TraceData): Record<string, unknown> {
    return {
      event_id: trace.traceId,
      user_id: trace.userId,
      event: 'ai_interaction',
      timestamp: new Date(trace.startTime).toISOString(),
      properties: {
        provider: trace.provider,
        conversation_id: trace.conversationId,
        latency_ms: trace.latencyMs,
        // Token usage
        ...(trace.tokens && {
          input_tokens: trace.tokens.input,
          output_tokens: trace.tokens.output,
          total_tokens: trace.tokens.total,
        }),
        // Error info
        ...(trace.error && { error: trace.error }),
        ...trace.properties,
      },
      ai_data: {
        model: trace.model,
        input: typeof trace.input === 'string' ? trace.input : JSON.stringify(trace.input),
        output: trace.output ? (typeof trace.output === 'string' ? trace.output : JSON.stringify(trace.output)) : undefined,
        convo_id: trace.conversationId,
      },
      // Include tool calls if present
      ...(trace.toolCalls && trace.toolCalls.length > 0 && {
        attachments: trace.toolCalls.map(tc => ({
          type: 'code',
          name: `tool:${tc.name}`,
          value: JSON.stringify({ arguments: tc.arguments, result: tc.result }),
          role: 'output',
          language: 'json',
        })),
      }),
    };
  }

  /**
   * Enqueue an event for sending
   */
  private enqueue(event: QueuedEvent): void {
    this.queue.push(event);

    if (this.config.debug) {
      console.log('[raindrop] Queued event:', event.type, event.data);
    }

    // Flush if queue is full
    if (this.queue.length >= this.MAX_QUEUE_SIZE) {
      this.flush();
      return;
    }

    // Schedule flush
    if (!this.flushTimeout) {
      this.flushTimeout = setTimeout(() => this.flush(), this.FLUSH_INTERVAL);
    }
  }

  /**
   * Flush all queued events
   */
  async flush(): Promise<void> {
    if (this.flushTimeout) {
      clearTimeout(this.flushTimeout);
      this.flushTimeout = null;
    }

    if (this.queue.length === 0) return;

    const events = [...this.queue];
    this.queue = [];

    // Group by type for batch sending
    const traces = events.filter(e => e.type === 'trace').map(e => e.data);
    const feedbacks = events.filter(e => e.type === 'feedback').map(e => e.data);
    const identifies = events.filter(e => e.type === 'identify').map(e => e.data);

    // Send in parallel, fire-and-forget
    const promises: Promise<void>[] = [];

    if (traces.length > 0) {
      promises.push(this.sendBatch('/events/track', traces));
    }
    if (feedbacks.length > 0) {
      promises.push(this.sendBatch('/signals/track', feedbacks));
    }
    for (const identify of identifies) {
      promises.push(this.sendSingle('/users/identify', identify));
    }

    // Wait but don't throw
    await Promise.allSettled(promises);
  }

  /**
   * Send a batch of events
   */
  private async sendBatch(endpoint: string, data: unknown[], retries = 0): Promise<void> {
    try {
      const response = await fetch(`${this.config.baseUrl}/v1${endpoint}`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${this.config.apiKey}`,
        },
        body: JSON.stringify(data),
      });

      if (!response.ok && retries < this.MAX_RETRIES) {
        if (this.config.debug) {
          console.warn(`[raindrop] Request failed (${response.status}), retrying...`);
        }
        await this.delay(Math.pow(2, retries) * 100);
        return this.sendBatch(endpoint, data, retries + 1);
      }

      if (this.config.debug && response.ok) {
        console.log(`[raindrop] Sent ${data.length} events to ${endpoint}`);
      }
    } catch (error) {
      if (retries < this.MAX_RETRIES) {
        await this.delay(Math.pow(2, retries) * 100);
        return this.sendBatch(endpoint, data, retries + 1);
      }
      if (this.config.debug) {
        console.warn('[raindrop] Failed to send events:', error);
      }
      // Fire-and-forget: swallow the error
    }
  }

  /**
   * Send a single event
   */
  private async sendSingle(endpoint: string, data: unknown, retries = 0): Promise<void> {
    try {
      const response = await fetch(`${this.config.baseUrl}/v1${endpoint}`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${this.config.apiKey}`,
        },
        body: JSON.stringify(data),
      });

      if (!response.ok && retries < this.MAX_RETRIES) {
        await this.delay(Math.pow(2, retries) * 100);
        return this.sendSingle(endpoint, data, retries + 1);
      }
    } catch (error) {
      if (retries < this.MAX_RETRIES) {
        await this.delay(Math.pow(2, retries) * 100);
        return this.sendSingle(endpoint, data, retries + 1);
      }
      if (this.config.debug) {
        console.warn('[raindrop] Failed to send event:', error);
      }
    }
  }

  private delay(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }

  /**
   * Close transport and flush remaining events
   */
  async close(): Promise<void> {
    await this.flush();
  }
}

File: python/src/rd_mini/transport.py
"""
HTTP Transport for Raindrop
Fire-and-forget with buffering and retry
"""

import atexit
import json
import sys
import threading
import time
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any, Literal

import httpx

from rd_mini.types import FeedbackOptions, SignalOptions, SpanData, TraceData, UserTraits

# SDK metadata - keep in sync with pyproject.toml
SDK_NAME = "rd-mini"
SDK_VERSION = "0.1.0"
MAX_EVENT_SIZE_BYTES = 1 * 1024 * 1024  # 1 MB


def get_context() -> dict[str, Any]:
    """Get SDK context metadata to include in events."""
    return {
        "library": {
            "name": SDK_NAME,
            "version": SDK_VERSION,
        },
        "metadata": {
            "pyVersion": f"v{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}",
        },
    }


def safe_json_dumps(value: Any) -> str:
    """Safely serialize a value to JSON, handling circular refs and errors."""
    try:
        return json.dumps(value, default=_safe_serializer)
    except (TypeError, ValueError, OverflowError):
        return str(value)


def _safe_serializer(obj: Any) -> Any:
    """Custom serializer for non-JSON-serializable types."""
    # Handle common non-serializable types
    if hasattr(obj, "__dict__"):
        return obj.__dict__
    if hasattr(obj, "isoformat"):  # datetime
        return obj.isoformat()
    if isinstance(obj, bytes):
        return obj.decode("utf-8", errors="replace")
    if isinstance(obj, set):
        return list(obj)
    return str(obj)


def to_api_string(value: Any) -> str | None:
    """Convert value to string for API, returning None for None/empty."""
    if value is None:
        return None
    if isinstance(value, str):
        return value
    return safe_json_dumps(value)


@dataclass
class QueuedEvent:
    """Event queued for sending."""

    type: Literal["trace", "feedback", "identify", "interaction"]
    data: dict[str, Any]
    timestamp: float


class Transport:
    """HTTP transport with batching and retry."""

    def __init__(
        self,
        api_key: str,
        base_url: str = "https://api.raindrop.ai",
        debug: bool = False,
        disabled: bool = False,
        flush_interval: float = 1.0,
        max_queue_size: int = 100,
        max_retries: int = 3,
    ):
        self.api_key = api_key
        self.base_url = base_url
        self.debug = debug
        self.disabled = disabled
        self.flush_interval = flush_interval
        self.max_queue_size = max_queue_size
        self.max_retries = max_retries

        self._queue: list[QueuedEvent] = []
        self._lock = threading.Lock()
        self._flush_timer: threading.Timer | None = None
        self._client = httpx.Client(timeout=30.0)
        self._closed = False

        # Register cleanup on exit
        atexit.register(self.close)

    def send_trace(self, trace: TraceData) -> None:
        """Send a trace event."""
        if self.disabled:
            return

        self._enqueue(
            QueuedEvent(
                type="trace",
                data=self._format_trace(trace),
                timestamp=time.time(),
            )
        )

    def send_feedback(self, trace_id: str, feedback: FeedbackOptions) -> None:
        """Send feedback/signal event."""
        if self.disabled:
            return

        # Determine signal name and sentiment
        if feedback.score is not None:
            signal_name = "positive" if feedback.score >= 0.5 else "negative"
            sentiment = "POSITIVE" if feedback.score >= 0.5 else "NEGATIVE"
        else:
            signal_name = feedback.type or "negative"
            sentiment = "POSITIVE" if feedback.type == "thumbs_up" else "NEGATIVE"

        data: dict[str, Any] = {
            "event_id": trace_id,
            "signal_name": signal_name,
            "sentiment": sentiment,
            "signal_type": feedback.signal_type,
            "timestamp": feedback.timestamp or datetime.now(timezone.utc).isoformat(),
            "properties": {
                "score": feedback.score,
                "comment": feedback.comment,
                **feedback.properties,
            },
        }

        if feedback.attachment_id:
            data["attachment_id"] = feedback.attachment_id

        self._enqueue(QueuedEvent(type="feedback", data=data, timestamp=time.time()))

    def send_signal(self, options: SignalOptions) -> None:
        """Send a signal with full options."""
        if self.disabled:
            return

        props: dict[str, Any] = {}
        if options.comment:
            props["comment"] = options.comment
        if options.after:
            props["after"] = options.after
        props.update(options.properties)

        data: dict[str, Any] = {
            "event_id": options.event_id,
            "signal_name": options.name,
            "signal_type": options.type,
            "sentiment": options.sentiment or "NEGATIVE",
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "properties": props,
        }

        if options.attachment_id:
            data["attachment_id"] = options.attachment_id

        self._enqueue(QueuedEvent(type="feedback", data=data, timestamp=time.time()))

    def send_identify(self, user_id: str, traits: UserTraits) -> None:
        """Send user identification."""
        if self.disabled:
            return

        self._enqueue(
            QueuedEvent(
                type="identify",
                data={"user_id": user_id, "traits": traits.to_dict()},
                timestamp=time.time(),
            )
        )

    def send_interaction(
        self,
        interaction_id: str,
        user_id: str | None,
        event: str,
        input_text: str | None,
        output: str | None,
        start_time: float,
        end_time: float,
        latency_ms: int,
        conversation_id: str | None,
        properties: dict[str, Any],
        error: str | None,
        spans: list[SpanData],
        attachments: list[dict[str, Any]] | None = None,
    ) -> None:
        """Send an interaction with nested spans."""
        if self.disabled:
            return

        # Convert spans to attachments
        span_attachments = []
        for span in spans:
            span_attachments.append(
                {
                    "type": "code",
                    "name": f"{span.type}:{span.name}",
                    "value": safe_json_dumps(
                        {
                            "spanId": span.span_id,
                            "input": span.input,
                            "output": span.output,
                            "latencyMs": span.latency_ms,
                            "error": span.error,
                            "properties": span.properties,
                        }
                    ),
                    "role": "output",
                    "language": "json",
                }
            )

        # Combine user attachments with span attachments
        all_attachments = (attachments or []) + span_attachments

        data: dict[str, Any] = {
            "event_id": interaction_id,
            "user_id": user_id,
            "event": event,
            "timestamp": datetime.fromtimestamp(start_time, tz=timezone.utc).isoformat(),
            "properties": {
                "$context": get_context(),
                "latency_ms": latency_ms,
                "span_count": len(spans),
                **({"error": error} if error else {}),
                **properties,
            },
            "ai_data": {
                "input": to_api_string(input_text),
                "output": to_api_string(output),
                "convo_id": conversation_id,
            },
        }

        if all_attachments:
            data["attachments"] = all_attachments

        self._enqueue(QueuedEvent(type="interaction", data=data, timestamp=time.time()))

    def _format_trace(self, trace: TraceData) -> dict[str, Any]:
        """Format trace data for API."""
        data: dict[str, Any] = {
            "event_id": trace.trace_id,
            "user_id": trace.user_id,
            "event": "ai_interaction",
            "timestamp": datetime.fromtimestamp(trace.start_time, tz=timezone.utc).isoformat(),
            "properties": {
                "$context": get_context(),
                "provider": trace.provider,
                "conversation_id": trace.conversation_id,
                "latency_ms": trace.latency_ms,
                **(
                    {
                        "input_tokens": trace.tokens.get("input"),
                        "output_tokens": trace.tokens.get("output"),
                        "total_tokens": trace.tokens.get("total"),
                    }
                    if trace.tokens
                    else {}
                ),
                **({"error": trace.error} if trace.error else {}),
                **trace.properties,
            },
            "ai_data": {
                "model": trace.model,
                "input": to_api_string(trace.input),
                "output": to_api_string(trace.output),
                "convo_id": trace.conversation_id,
            },
        }

        # Add tool calls as attachments
        if trace.tool_calls:
            data["attachments"] = [
                {
                    "type": "code",
                    "name": f"tool:{tc.get('name', 'unknown')}",
                    "value": safe_json_dumps(
                        {"arguments": tc.get("arguments"), "result": tc.get("result")}
                    ),
                    "role": "output",
                    "language": "json",
                }
                for tc in trace.tool_calls
            ]

        return data

    def _enqueue(self, event: QueuedEvent) -> None:
        """Add event to queue and schedule flush."""
        # Check event size
        try:
            event_size = len(json.dumps(event.data).encode("utf-8"))
            if event_size > MAX_EVENT_SIZE_BYTES:
                if self.debug:
                    print(
                        f"[raindrop] Event exceeds 1MB limit ({event_size / 1024 / 1024:.2f}MB), skipping"
                    )
                return
        except (TypeError, ValueError):
            # If we can't serialize, let it through and let the API handle it
            pass

        with self._lock:
            # Check buffer capacity
            if len(self._queue) >= self.max_queue_size:
                if self.debug:
                    print("[raindrop] Buffer full, discarding oldest event")
                self._queue.pop(0)  # Remove oldest event
            elif len(self._queue) >= int(self.max_queue_size * 0.8):
                if self.debug:
                    print(
                        f"[raindrop] Buffer at {round(len(self._queue) / self.max_queue_size * 100)}% capacity"
                    )

            self._queue.append(event)

            if self.debug:
                print(f"[raindrop] Queued event: {event.type} {event.data}")

            # Schedule flush if not already scheduled
            if self._flush_timer is None and not self._closed:
                self._flush_timer = threading.Timer(self.flush_interval, self._flush_now)
                self._flush_timer.daemon = True
                self._flush_timer.start()

    def _flush_now(self) -> None:
        """Flush all queued events."""
        with self._lock:
            if self._flush_timer:
                self._flush_timer.cancel()
                self._flush_timer = None

            if not self._queue:
                return

            events = self._queue[:]
            self._queue = []

        # Group by type
        traces = [e.data for e in events if e.type in ("trace", "interaction")]
        feedbacks = [e.data for e in events if e.type == "feedback"]
        identifies = [e.data for e in events if e.type == "identify"]

        # Send in parallel (fire-and-forget)
        if traces:
            self._send_batch("/events/track", traces)
        if feedbacks:
            self._send_batch("/signals/track", feedbacks)
        for identify in identifies:
            self._send_single("/users/identify", identify)

    def _send_batch(self, endpoint: str, data: list[dict[str, Any]], retries: int = 0) -> None:
        """Send a batch of events."""
        try:
            response = self._client.post(
                f"{self.base_url}/v1{endpoint}",
                json=data,
                headers={
                    "Content-Type": "application/json",
                    "Authorization": f"Bearer {self.api_key}",
                },
            )

            if not response.is_success and retries < self.max_retries:
                if self.debug:
                    print(f"[raindrop] Request failed ({response.status_code}), retrying...")
                time.sleep(0.1 * (2**retries))
                return self._send_batch(endpoint, data, retries + 1)

            if self.debug and response.is_success:
                print(f"[raindrop] Sent {len(data)} events to {endpoint}")

        except Exception as e:
            if retries < self.max_retries:
                time.sleep(0.1 * (2**retries))
                return self._send_batch(endpoint, data, retries + 1)
            if self.debug:
                print(f"[raindrop] Failed to send events: {e}")

    def _send_single(self, endpoint: str, data: dict[str, Any], retries: int = 0) -> None:
        """Send a single event."""
        try:
            response = self._client.post(
                f"{self.base_url}/v1{endpoint}",
                json=data,
                headers={
                    "Content-Type": "application/json",
                    "Authorization": f"Bearer {self.api_key}",
                },
            )

            if not response.is_success and retries < self.max_retries:
                time.sleep(0.1 * (2**retries))
                return self._send_single(endpoint, data, retries + 1)

        except Exception as e:
            if retries < self.max_retries:
                time.sleep(0.1 * (2**retries))
                return self._send_single(endpoint, data, retries + 1)
            if self.debug:
                print(f"[raindrop] Failed to send event: {e}")

    def flush(self) -> None:
        """Manually flush all pending events."""
        self._flush_now()

    def close(self) -> None:
        """Close transport and flush remaining events."""
        self._closed = True
        self._flush_now()
        self._client.close()
